### ISSUE TITLE
After selecting item on Typeahead, onChange callback triggered twice.

### ISSUE THREAD

**Issue by vudodov** [ericgio/react-bootstrap-typeahead#320]
### Version
2.4.0
### Steps to reproduce
Add `Typeahead` component which looks like this
```
<Typeahead
  onChange={(selected) => {
    this.setState({selected});
  }}
  options={['one', 'two']}
  selected={this.state.selected}
/>
```
Breakpoint inside on `onChange` 
Select any item
### Expected Behavior
`onChange` callback hitted **once**
### Actual Behavior
`onChange` callback hitted **twice**

### Notes
It looks like first call is comming from `typeaheadInputContainer.js`. Whenever user selects item, `_handleChange` for `Input` got executed, which triggers `this.props.onChange(...)`.
Second call is comming from `typeaheadContainer.js` because `selected` prop is updated, which triggers `componentWillReceiveProps(...)` -> `this._updateText(...)` -> `this.props.onInputChange(...)` -> `this.props.onChange()`

**Comment by bzashev** [comment]
I have the same problem too

**Comment by ericgio** [comment]
Thanks for the report. Sounds like the behavior described in #305. Seems like I need to rethink a bit how and where `onChange` and `onInputChange` are being called.

**Comment by vudodov** [comment]
I'm agree with @coolkev [his last comment](https://github.com/ericgio/react-bootstrap-typeahead/issues/305#id=issuecomment-365415298) in #305 

> onChange events should really not be raised in the componentWillReceiveProps event

**Comment by ericgio** [comment]
This is being fixed in v3.0.0

**Comment by ericgio** [comment]
This should now be fixed with the release of v3.0.0

**Comment by bzashev** [comment]
For me the problem persists in v.3.0.1. 
When I click with the mouse on a result in the type ahead dropdown all is OK, but if I use Enter to select it, there is a double entry

**Comment by ericgio** [comment]
@bzashev: Thanks for the report. I'm unable to repro the issue in the examples, based on your description. Could you please provide a code example or test case? Thanks.

**Comment by bzashev** [comment]
Hey, the way to test this is the above example but add  selectHintOnEnter={true}. Then try to use only the keyboard to select an item from the dropdown. This will trigger the onChange twice.

**Comment by ericgio** [comment]
Ok, I can repro the issue now, thanks for clarifying.

**Comment by ericgio** [comment]
@bzashev: Thanks again for the report, the issue should be fixed in v3.0.2 (2f0cec9861378955672b32736d35fed36c31cf6d). I added some tests as well to help prevent regressions.

**Comment by bzashev** [comment]
Thanks a lot @ericgio All the best. You can close this, I will reopen if I have problems

**Comment by bzashev** [comment]
I confirm it works now

**Comment by bzashev** [comment]
I have to take my words back.
I updated to 3.0.3 and there is a related case when it does double entry again. This when exactly the "selectHintOnEnter" is triggered.
You can simulate this by loading the example and only using the keyboard select an item from the dropdown. Then press backspace so the "selectHintOnEnter" option is triggered (the last character in the item is grayed, from which I suppose the control is in a hint mode). You get a double entry.

There is an improvement though as before I got double entry even if I git enter on a properly selected element.

**Comment by ericgio** [comment]
@bzashev: I'm not able to reproduce a double change event based on what you're describing.

> press backspace so the "selectHintOnEnter" option is triggered

Not sure exactly what you mean by this. If you've made a selection and hit backspace, then the selection will be cleared, triggering another change event. That's expected.

**Comment by bzashev** [comment]
Let me try to explain it better. 
When the dropdown with the possible results appear bellow the input, you can use the arrows to navigate through them. This hovers the item in the dropdown but also changes the value in the input to correspond with the hovered item. In this way, only by using the down arrow I select the first or the second item in the dropdown(eg. "one"). This makes in the input the word "one" to be presented. Then without pressing anything else, if I press the backspace button there will be a change in the input - the letters "on" will be in black, the letter "e" will be grayed out (you need to have "selectHintOnEnter" enabled. Then if I press immediately after the backspace the key Enter - there is a double selection.
In short here are the exact steps:
1. Click in the input. Dropdown appears with two items - "one" and "two"
2. Press the down arrow
3. Press Enter

Hope this helps

**Comment by ericgio** [comment]
@bzashev: I'm sorry, but I still can't reproduce a double change event based on what you're describing. I'm also a bit confused since you describe hitting backspace, but don't include that in the numbered repro steps.

I wrote a test based on the steps you describe above and it passes as expected:
```jsx
it('only triggers one onChange event', () => {
  let selected = [];
  const onChange = sinon.spy((s) => selected = s);

  const wrapper = mount(
    <Typeahead
      onChange={onChange}
      options={['one', 'two']}
      selected={selected}
      selectHintOnEnter
    />
  );

  focus(wrapper);
  keyDown(wrapper, DOWN);
  keyDown(wrapper, RETURN);

  expect(selected.length).to.equal(1);
  expect(onChange.calledOnce).to.equal(true);
});
```

**Comment by bzashev** [comment]
Sorry for the late reply. Two things which may not be of consequence but I noticed. 
1. should selectHintOnEnter should equal to true?
2. between the Down and Return keydowns, shouldn't you have a Backspace?

**Comment by bzashev** [comment]
Hey,
you are really fast. I updated the comment, right you were, it was missing in the numbered list.
I will look deep into the example I have and what influences it, as it should be some more complex thing with properties or probably browser too.
Will feedback soon

**Comment by GunaySeyidahmed** [comment]
sde
